File: 2016/01/01.py
def generator(input: str): 
    x, y : int = 0 , 0
    ways = [(1,0) , (0, -1), (-1, 0), (0,1)]
    path = [(x,y)]
    dir = 0
    for step in input.split(', '):
        match step[0]:
            case 'L': dir = (dir - 1) % 4
            case 'R': dir = (dir + 1) % 4
        dx, dy = ways[dir]
        for step in range(int(step[1:])):
            path.append((x:= x + dx, y:= y + dy))
    return path

def part_1(path):
    x, y = path.pop()
    return abs(x) + abs(y)

def part_2(path): 
    path = path[::-1]
    x, y = path.pop()
    visited = {(x, y)}
    while (point := path.pop()) not in visited: 
        visited.add(point) 
    return abs(point[0]) + abs(point[1])

File: 2016/01/01.rs
aoc2016::main!();

fn generator(input: &str) -> Vec<(i32,i32)> {
    let mut dir = (1,0);
    let mut pos = (0,0);
    input
        .split(", ")
        .map(move |info| {
            let turn = info.clone().chars().nth(0).unwrap() ;
            let l = info[1..].parse::<i32>().unwrap() ;
            match turn {
                'R' => dir = turn_right(dir),
                'L' => dir = turn_left(dir),
                _ => unreachable!()} ;
                (dir, l) })
        .map(|(dir, l)| {(1..l+1)
            .scan(pos, |_,_| {pos = (pos.0 + dir.0, pos.1 + dir.1);
                Some(pos)})
            .collect_vec()})
        .flatten()
        .collect_vec()
}


fn part_1(input: Vec<(i32,i32)>) -> i32 {
    input
        .iter()
        .last()
        .map(|d| d.0.abs() + d.1.abs())
        .unwrap()
}

fn part_2(input: Vec<(i32,i32)>) -> i32 {
    input
        .iter()
        .enumerate()
        .find(|&(n,pos)| input.iter().take(n).contains(pos))
        .map(|(_,d)| d.0.abs() + d.1.abs())
        .unwrap()
}


fn turn_left(input: (i32, i32)) -> (i32, i32){
    match input {
        (1,0) => (0,1),
        (0,1) => (-1,0),
        (-1,0) => (0,-1),
        (0,-1) => (1,0),
        _ => unreachable!()
    }
}

fn turn_right(input: (i32, i32)) -> (i32, i32){
    match input {
        (1,0) => (0,-1),
        (0,-1) => (-1,0),
        (-1,0) => (0,1),
        (0,1) => (1,0),
        _ => unreachable!()
    }
}